src/pages/academy/adminPanel/subcomponents/NotificationConfigPanel.tsx

```diff
@@ -0,0 +1,224 @@
+    const temp = [...notificationConfig.current];
+
+    temp[index] = {
+      ...temp[index],
+      timeOptions: value
+    };
+    setNotificationConfig(temp);
+    gridApi.current?.getDisplayedRowAtIndex(index)?.setDataValue('timeOptions', value);
+    setHasChangesNotificationConfig(true);
+  };
+
+  const assessmentTypeFormatter: ValueFormatterFunc<NotificationConfiguration> = params => {
+    return params.data!.assessmentConfig?.type || '-';
+  };
+
+  const recipientFormatter: ValueFormatterFunc<NotificationConfiguration> = params => {
+    return params.data!.notificationType.forStaff ? 'Staff' : 'Student';
+  };
+
+  const columnDefs = [
+    {
+      headerName: 'Notification Type',
+      field: 'notificationType.name',
+      rowDrag: true
+    },
+    {
+      headerName: 'Assessment Type',
+      field: 'assessmentConfig.type',
+      valueFormatter: assessmentTypeFormatter
+    },
+    {
+      headerName: 'Recipients',
+      field: 'notificationType.forStaff',
+      valueFormatter: recipientFormatter
+    },
+    // TODO: Extension for notification configs
+    // {
+    //   headerName: 'View Email Template',
+    //   field: 'notificationType.name'
+    // },
+    // {
+    //   headerName: 'Past 30 Days',
+    //   field: 'notificationType.id'
+    // },
+    {
+      headerName: 'Reminder Time Options (hours)',
+      field: 'timeOptions',
+      cellRendererFramework: TimeOptionCell,
+      cellRendererParams: {
+        setStateHandler: setTimeOptions,
+        setDelete: addTimeOptionsToDelete,
+        field: 'timeOptions'
+      }
+    },
+    {
+      headerName: 'Default Reminder (hours)',
+      field: 'timeOptions',
+      cellRendererFramework: SelectCell,
+      cellRendererParams: {
+        setStateHandler: setTimeOptions,
+        field: 'timeOptions'
+      }
+    },
+    {
+      headerName: 'Enabled',
+      field: 'isEnabled',
+      cellRendererFramework: BooleanCell,
+      cellRendererParams: {
+        setStateHandler: setIsEnabled,
+        field: 'isEnabled'
+      }
+    }
+  ];
+
+  const defaultColumnDefs = {
+    filter: false,
+    resizable: true,
+    sortable: false
+  };
+
+  const onGridReady = (params: GridReadyEvent) => {
+    gridApi.current = params.api;
+    params.api.sizeColumnsToFit();
+  };
+
+  // Handler to submit changes to Notification Configration to the backend.
+  // Changes made to users, Course Configration and Assessment Configuration are handled separately.
+  const submitHandler = () => {
+    if (hasChangesNotificationConfig) {
+      setHasChangesNotificationConfig(false);
+      const allTimeOptions = notificationConfig.current.reduce((acc, curr) => {
+        const timeOptions = curr.timeOptions.map(timeOption => {
+          return {
+            ...timeOption,
+            notificationConfigId: curr.id
+          };
+        });
+        return acc.concat(timeOptions);
+      }, [] as TimeOption[]);
```

Comment:
This feels like a very weird way to achieve what you are trying to do. Ditto on typecast, I personally feel it can be avoided. Let's not overcomplicate this code.


```suggestion
      const allTimeOptions: TimeOption[] = [];
      notificationConfig.current?.forEach((curr) => {
        const timeOptions = curr.timeOptions.map(timeOption => {
          return {
            ...timeOption,
            notificationConfigId: curr.id
          };
        });
        allTimeOptions.push(...timeOptions);
      });
```

Reply:
The above suggestion is also more type-safe as it guards against `notificationConfig.current` being undefined.